### PR TITLE
BUG: Fix DataFrame.groupby().apply() for NaN groups with dropna=False 

### LINKED ISSUE #35889: BUG: groupby dropna=False with nan value in groupby causes ValueError when apply()
- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the latest version of pandas, version `1.1.1`

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python
import pandas as pd
import numpy as np

df = pd.DataFrame({
    'groups': ['a', 'a', 'b', np.nan],
    'tonnages': [10, 10, 20, 30]
})
dfg = df.groupby('groups', dropna=False)
rv = dfg.apply(lambda grp: pd.DataFrame({'values': list(range(len(grp)))}))
```

#### Problem description

`ValueError` raised on `dfg.apply(lambda grp: ...)`; with the following stacktrace:
<details>

```python
../../env3.8/lib/python3.8/site-packages/pandas/core/groupby/groupby.py:859: in apply
    result = self._python_apply_general(f, self._selected_obj)
../../env3.8/lib/python3.8/site-packages/pandas/core/groupby/groupby.py:894: in _python_apply_general
    return self._wrap_applied_output(
../../env3.8/lib/python3.8/site-packages/pandas/core/groupby/generic.py:1230: in _wrap_applied_output
    return self._concat_objects(keys, values, not_indexed_same=not_indexed_same)
../../env3.8/lib/python3.8/site-packages/pandas/core/groupby/groupby.py:1145: in _concat_objects
    result = concat(
../../env3.8/lib/python3.8/site-packages/pandas/core/reshape/concat.py:274: in concat
    op = _Concatenator(
../../env3.8/lib/python3.8/site-packages/pandas/core/reshape/concat.py:454: in __init__
    self.new_axes = self._get_new_axes()
../../env3.8/lib/python3.8/site-packages/pandas/core/reshape/concat.py:519: in _get_new_axes
    return [
../../env3.8/lib/python3.8/site-packages/pandas/core/reshape/concat.py:520: in <listcomp>
    self._get_concat_axis() if i == self.bm_axis else self._get_comb_axis(i)
../../env3.8/lib/python3.8/site-packages/pandas/core/reshape/concat.py:576: in _get_concat_axis
    concat_axis = _make_concat_multiindex(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

indexes = [RangeIndex(start=0, stop=2, step=1), RangeIndex(start=0, stop=1, step=1), RangeIndex(start=0, stop=1, step=1)]
keys = Index(['a', 'b', nan], dtype='object', name='groups')
levels = [Index(['a', 'b', nan], dtype='object', name='groups')]
names = ['groups']

    def _make_concat_multiindex(indexes, keys, levels=None, names=None) -> MultiIndex:
    
        if (levels is None and isinstance(keys[0], tuple)) or (
            levels is not None and len(levels) > 1
        ):
            zipped = list(zip(*keys))
            if names is None:
                names = [None] * len(zipped)
    
            if levels is None:
                _, levels = factorize_from_iterables(zipped)
            else:
                levels = [ensure_index(x) for x in levels]
        else:
            zipped = [keys]
            if names is None:
                names = [None]
    
            if levels is None:
                levels = [ensure_index(keys)]
            else:
                levels = [ensure_index(x) for x in levels]
    
        if not all_indexes_same(indexes):
            codes_list = []
    
            # things are potentially different sizes, so compute the exact codes
            # for each level and pass those to MultiIndex.from_arrays
    
            for hlevel, level in zip(zipped, levels):
                to_concat = []
                for key, index in zip(hlevel, indexes):
                    mask = level == key
                    if not mask.any():
>                       raise ValueError(f"Key {key} not in level {level}")
E                       ValueError: Key nan not in level Index(['a', 'b', nan], dtype='object', name='groups')

../../env3.8/lib/python3.8/site-packages/pandas/core/reshape/concat.py:631: ValueError
```

</details>

#### Expected Output

No error should be raised. With the above, if i omit the `nan`:

```python
df = pd.DataFrame({
    'groups': ['a', 'a', 'b'],
    'tonnages': [10, 10, 20]
})
dfg = df.groupby('groups', dropna=False)
rv = dfg.apply(lambda grp: pd.DataFrame({'values': list(range(len(grp)))}))
```
Then it works successfully with `rv` being:

```python
          values
groups          
a      0       0
       1       1
b      0       0
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : f2ca0a2665b2d169c97de87b8e778dbed86aea07
python           : 3.8.2.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 18.7.0
Version          : Darwin Kernel Version 18.7.0: Thu Jun 18 20:50:10 PDT 2020; root:xnu-4903.278.43~1/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_AU.UTF-8
LOCALE           : en_AU.UTF-8

pandas           : 1.1.1
numpy            : 1.18.4
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.1.1
setuptools       : 46.4.0
Cython           : 0.29.17
pytest           : 5.1.1
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : 0.9.6
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : 2.8.3 (dt dec pq3 ext lo64)
jinja2           : 2.11.2
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : 1.3.1
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : 2.7.1
odfpy            : None
openpyxl         : 1.8.6
pandas_gbq       : None
pyarrow          : None
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : 1.3.12
tables           : 3.6.1
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by cwkwong**
- [X] closes #35889 
- [X] tests added / passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry in `v1.1.2.rst`

### CODE REVIEW

File: pandas/tests/groupby/test_groupby_dropna.py [hunk 1]
```diff
@@ -276,3 +276,60 @@ def test_groupby_dropna_datetime_like_data(
...
+)
+def test_groupby_dropna_multi_index_dataframe_apply(
+    dropna, df_cols_in, df_cols_out, levels
+):
+    # GH 35889
+    # `groupby` with `dropna=False` and `apply` returning DataFrame of different
+    # sizes raises error if grouped column has nan values.
+
+    df = pd.DataFrame(df_cols_in)
+    dfg = df.groupby("groups", dropna=dropna)
```
Comment (arw2019) [inline pandas/tests/groupby/test_groupby_dropna.py:324]: nitpick: call this `gb`

File: pandas/tests/groupby/test_groupby_dropna.py [hunk 2]
```diff
@@ -276,3 +276,60 @@ def test_groupby_dropna_datetime_like_data(
...
+    rv = dfg.apply(lambda grp: pd.DataFrame({"values": list(range(len(grp)))}))
```
Comment (arw2019) [inline pandas/tests/groupby/test_groupby_dropna.py:325]: another nitpick: call this `result`

File: pandas/tests/groupby/test_groupby_dropna.py [hunk 3]
```diff
@@ -276,3 +276,60 @@ def test_groupby_dropna_datetime_like_data(
...
+            # no nan in "groups"; dropna=True|False should be same.
+            True,
+            {"groups": ["a", "a", "b", "c"], "values": [10, 10, 20, 30]},
+            {"values": [0, 1, 0, 0]},
+            None,
+            id="dropna_true_no_nan",
+        ),
+    ],
+)
+def test_groupby_dropna_multi_index_dataframe_apply(
```
Comment (arw2019) [inline pandas/tests/groupby/test_groupby_dropna.py:316]: I'd do `test_groupby_apply_with_dropna_for_multiindex`

File: pandas/tests/groupby/test_groupby_dropna.py [hunk 4]
```diff
@@ -276,3 +276,60 @@ def test_groupby_dropna_datetime_like_data(
...
+    dropna, df_cols_in, df_cols_out, levels
+):
+    # GH 35889
+    # `groupby` with `dropna=False` and `apply` returning DataFrame of different
```
Comment (arw2019) [inline pandas/tests/groupby/test_groupby_dropna.py:320]: nit-pick: probably don't need this, GH issue and test name are enough info

File: pandas/tests/groupby/test_groupby_dropna.py (hunk 2 above)
Comment (arw2019) [reply pandas/tests/groupby/test_groupby_dropna.py:325]: also you can feed `range` directly to the DataFrame constructor, no need to convert to list

File: pandas/tests/groupby/test_groupby_dropna.py [hunk 5]
```diff
@@ -276,3 +276,60 @@ def test_groupby_dropna_datetime_like_data(
...
+    # sizes raises error if grouped column has nan values.
+
+    df = pd.DataFrame(df_cols_in)
+    dfg = df.groupby("groups", dropna=dropna)
+    rv = dfg.apply(lambda grp: pd.DataFrame({"values": list(range(len(grp)))}))
+
+    tuples = tuple(zip(df_cols_in["groups"], df_cols_out["values"]))
```
Comment (arw2019) [inline pandas/tests/groupby/test_groupby_dropna.py:327]: nitpick: I'd call this `mi_tuples`

File: pandas/tests/groupby/test_groupby_dropna.py [hunk 6]
```diff
@@ -276,3 +276,60 @@ def test_groupby_dropna_datetime_like_data(
...
+            True,
+            {"groups": ["a", "a", "b", "c"], "values": [10, 10, 20, 30]},
+            {"values": [0, 1, 0, 0]},
+            None,
+            id="dropna_true_no_nan",
+        ),
+    ],
+)
+def test_groupby_dropna_multi_index_dataframe_apply(
+    dropna, df_cols_in, df_cols_out, levels
```
Comment (arw2019) [inline pandas/tests/groupby/test_groupby_dropna.py:317]: I'd rename `df_cols_in` &#8594; `data`  and `df_cols_out` &#8594; `selected_data`

File: doc/source/whatsnew/v1.1.2.rst [hunk 7]
```diff
@@ -29,6 +29,7 @@ Bug fixes
 - Bug in :class:`Series` constructor raising a ``TypeError`` when constructing sparse datetime64 dtypes (:issue:`35762`)
 - Bug in :meth:`DataFrame.apply` with ``result_type="reduce"`` returning with incorrect index (:issue:`35683`)
 - Bug in :meth:`DateTimeIndex.format` and :meth:`PeriodIndex.format` with ``name=True`` setting the first item to ``"None"`` where it should bw ``""`` (:issue:`35712`)
+- Bug in :meth:`DataFrame.apply` on :meth:`DataFrame.groupby`, ``dropna=False`` and ``np.nan`` group(s) (:issue:`35889`)
```
Comment (arw2019) [inline doc/source/whatsnew/v1.1.2.rst:32]: Bug in :meth:`DataFrame.groupby(...).apply(...)` raising error with ``np.nan`` group(s) when `dropna=False`

**Review by arw2019 — CHANGES_REQUESTED** [review]
thanks @cwkwong for a nice PR!

Some largely cosmetic comments, otherwise lgtm

**Review by arw2019 — APPROVED** [review]

**Comment by arw2019** [comment]
@cwkwong might want to have a more descriptive title `BUG: Fix DataFrame.groupby().apply() for NaN groups with dropna=False`

**Comment by cwkwong** [comment]
Thanks @arw2019 for approving the changes. What's the next step to hopefully make this as part of the `1.1.2` release?

File: doc/source/whatsnew/v1.1.2.rst [hunk 8]
```diff
@@ -29,6 +29,7 @@ Bug fixes
 - Bug in :class:`Series` constructor raising a ``TypeError`` when constructing sparse datetime64 dtypes (:issue:`35762`)
 - Bug in :meth:`DataFrame.apply` with ``result_type="reduce"`` returning with incorrect index (:issue:`35683`)
 - Bug in :meth:`DateTimeIndex.format` and :meth:`PeriodIndex.format` with ``name=True`` setting the first item to ``"None"`` where it should bw ``""`` (:issue:`35712`)
+- Bug in :meth:`DataFrame.groupby(...).apply(...)` raising error with ``np.nan`` group(s) when ``dropna=False`` (:issue:`35889`)
```
Comment (charlesdong1991) [inline doc/source/whatsnew/v1.1.2.rst:32]:
```suggestion
- Bug in :meth:`DataFrameGroupBy.apply` raising error with ``np.nan`` group(s) when ``dropna=False`` (:issue:`35889`)
```

File: pandas/core/reshape/concat.py [hunk 9]
```diff
@@ -619,17 +620,16 @@ def _make_concat_multiindex(indexes, keys, levels=None, names=None) -> MultiInde
         codes_list = []
 
         # things are potentially different sizes, so compute the exact codes
-        # for each level and pass those to MultiIndex.from_arrays
-
+        # for each level and pass those to MultiIndex.from_arrays.
```
Comment (charlesdong1991) [inline pandas/core/reshape/concat.py:623]: could you please revert the change here? ideally only change the code where needs to be changed

File: pandas/core/reshape/concat.py [hunk 10]
```diff
@@ -619,17 +620,16 @@ def _make_concat_multiindex(indexes, keys, levels=None, names=None) -> MultiInde
...
             for key, index in zip(hlevel, indexes):
-                mask = level == key
+                mask = (isna(level) & isna(key)) | (level == key)
                 if not mask.any():
                     raise ValueError(f"Key {key} not in level {level}")
-                i = np.nonzero(level == key)[0][0]
-
+                i = np.nonzero(mask)[0][0]
                 to_concat.append(np.repeat(i, len(index)))
+
```
Comment (charlesdong1991) [inline pandas/core/reshape/concat.py:632]: same here?

File: pandas/core/reshape/concat.py [hunk 11]
```diff
@@ -619,17 +620,16 @@ def _make_concat_multiindex(indexes, keys, levels=None, names=None) -> MultiInde
...
 
         # things are potentially different sizes, so compute the exact codes
-        # for each level and pass those to MultiIndex.from_arrays
-
+        # for each level and pass those to MultiIndex.from_arrays.
         for hlevel, level in zip(zipped, levels):
             to_concat = []
             for key, index in zip(hlevel, indexes):
-                mask = level == key
+                mask = (isna(level) & isna(key)) | (level == key)
```
Comment (charlesdong1991) [inline pandas/core/reshape/concat.py:627]: could you pls add a comment above this line to describe the `mask` behaviour?

**Review by charlesdong1991 — CHANGES_REQUESTED** [review]
thanks for the PR @cwkwong 

This should be moved to whatsnew 1.2.0 note. 1.1.2 mainly focuses on regression issue.

**Comment by jreback** [comment]
thanks @cwkwong very nice

**Pull request merged**